### PR TITLE
EnterContextualReflection handle null

### LINKED ISSUE #23952: AssemblyLoadContext.EnterContextualReflection could produce a null reference exception
During nullable annotations I noticed that when calling `AssemblyLoadContext.EnterContextualReflection(Assembly)` with an implementation of `Assembly` that it's base type is not `RuntimeAssembly` it would produce a null reference exception. 

We call `GetLoadContext` which will return null if the `Assembly` is not a `RuntimeAssembly` -- then we try to dereference it without checking for null.

https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.CoreCLR.cs#L257

https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs#L513

cc: @stephentoub @jkotas @sdmaclea

### PR DESCRIPTION
**Pull request by sdmaclea**
Fixes #23952 

/cc @safern @jkotas @stephentoub

### CODE REVIEW

File: src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs [hunk 1]
```diff
@@ -507,11 +507,11 @@ public ContextualReflectionScope EnterContextualReflection()
...
         /// </remarks>
-        public static ContextualReflectionScope EnterContextualReflection(Assembly activating)
+        public static ContextualReflectionScope EnterContextualReflection(Assembly? activating)
         {
-            return activating != null ?
-                GetLoadContext(activating).EnterContextualReflection() :
-                new ContextualReflectionScope(null);
+            AssemblyLoadContext? assemblyLoadContext = (activating != null) ?
+                GetLoadContext(activating) : null;
+            return assemblyLoadContext?.EnterContextualReflection() ?? ContextualReflectionScope(null);
```
Comment (safern) [inline src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs]: Wouldn't it be better produced code to do

```cs
if (assemblyLoadContext == null)
     return assemblyLoadContext.EnterContextualReflection();

return ContextualReflectionScope(null);
```

Comment (sdmaclea) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs]: I am not a codegen expert.  It maybe more readable.

File: src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs [hunk 2]
```diff
@@ -509,9 +509,13 @@ public ContextualReflectionScope EnterContextualReflection()
         /// </remarks>
         public static ContextualReflectionScope EnterContextualReflection(Assembly activating)
         {
-            return activating != null ?
-                GetLoadContext(activating).EnterContextualReflection() :
-                new ContextualReflectionScope(null);
+            AssemblyLoadContext? assemblyLoadContext = (activating != null) ?
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs]: Would it be better to throw here?

- `ArgumentNullException` for null activating assembly
- `ArgumentException(SR.Arg_MustBeType, nameof(activating));` for non-runtime assembly.

Comment (sdmaclea) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs]: >  `ArgumentException(SR.Arg_MustBeType, nameof(activating));` for non-runtime assembly.

Sure, but should be `SR.Arg_MustBeRuntimeAssembly` which needs to be added...
 `ArgumentException(SR.Arg_MustBeRuntimeAssembly, nameof(activating));` for non-runtime assembly.

Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs]: Yes, that's what I meant...

**Pull request merged**